interfaces/cython/cantera/composite.py

```diff
@@ -528,32 +528,46 @@ def __init__(self, phase, shape=(0,), states=None, extra=None, meta=None):
         reserved = self.__dir__()
 
         self._extra = OrderedDict()
+
         if isinstance(extra, dict):
             for name, v in extra.items():
                 if name in reserved:
                     raise ValueError(
                         "Unable to create extra column '{}': name is already "
                         "used by SolutionArray objects.".format(name))
                 if not np.shape(v):
-                    self._extra[name] = [v]*self._shape[0]
-                elif len(v) == self._shape[0]:
-                    self._extra[name] = list(v)
+                    self._extra[name] = np.full(self._shape, v)
+                elif np.array(v).shape == self._shape:
+                    self._extra[name] = np.array(v)
                 else:
-                    raise ValueError("Unable to map extra SolutionArray"
-                                     "input for named {!r}".format(name))
+                    raise ValueError("Unable to map extra SolutionArray "
+                                     "input named {!r}".format(name))
+        elif extra is not None:
+            if isinstance(extra, np.ndarray):
+                extra = extra.flatten()
+            elif isinstance(extra, str):
+                extra = [extra]
 
-        elif extra and self._shape == (0,):
-            for name in extra:
+            try:
+                iter_extra = iter(extra)
+            except TypeError :
+                raise ValueError(
+                    "Extra properties can be created by passing an iterable "
+                    "of names for the properties. If you want to supply initial "
+                    "values for the properties, use a dictionary whose keys are "
+                    "the names of the properties and values are the initial "
+                    "values.") from None
+
+            for name in iter_extra:
+                if not isinstance(name, str):
+                    raise TypeError(
+                        "Unable to create extra column, passed value '{!r}' "
+                        "is not a string".format(name))
                 if name in reserved:
                     raise ValueError(
                         "Unable to create extra column '{}': name is already "
                         "used by SolutionArray objects.".format(name))
-                self._extra[name] = []
-
-        elif extra:
-            raise ValueError("Initial values for extra properties must be "
-                             "supplied in a dict if the SolutionArray is not "
-                             "initially empty")
+                self._extra[name] = np.empty(self._shape)
```

Comment:
@speth ... I’d ask to consider retaining this behavior. 

Think of a scenario where `extra` is calculated based on the content of the `SolutionArray` itself (which typically receives the actual value using some setter only after it is instantiated, so the content of extra is not known a priori). For this case, you’d have to provide dummy values to create `extra` columns, which would be subsequently overwritten. I believe that automatically initializing values as zeros would be a lot more convenient; columns that contain strings are rare, so having to provide a value there would be ok.

Reply:
@ischoegl I originally had the same objection as you, but I considered it for a while and I ended up agreeing with @speth. There are two main reasons for this, as I see it:

1. It is trivially easy to specify a dummy value, since you do not need to specify the shape. For instance:

```python
>>> arr = ct.SolutionArray(gas, (3, 10), extra={"prop": 0, "prop2": 0.0})
>>> arr.prop
array([[0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0],
       [0, 0, 0, 0, 0, 0, 0, 0, 0, 0]])
>>> arr.prop2
array([[0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.],
       [0., 0., 0., 0., 0., 0., 0., 0., 0., 0.]])
>>> # This is not allowed right now, but would be almost the same, except the dtype
>>> arr = ct.SolutionArray(gas, (3, 10), extra=["prop", "prop2"])
```

Since I've used `np.full()` to fill the initial value, there's no need to give an array as the initial value.

2. It isn't clear what `dtype` should be used - it is not only strings, but whether one wants to use floats vs. ints or any other dtype as well. I suppose floats could be used in place of ints, but there are legitimate use cases for the int dtype, for instance, as a categorical or level variable. I think it would also not be clear why one would have to specify an initial value for some dtypes but not for others.

Reply:
@bryanwweber ... I didn’t think of the scalar route, so your argument is convincing.

Reply:
The availability of the scalar initializer is a nice compromise for the case where the value is going to be calculated anyway.

If I'm reading this block of code correctly now, the only time that this line can be called is if `self._shape` is `(0,)`, right? It might be more clear if the value was just set to `self._extra[name] = np.empty(0)` then.

Reply:
Yes, that's correct, it only runs if `self._shape == (0,)`. I'm not sure it's clearer to just pass `0` though. Shapes of NumPy arrays are always returned as tuples, and just writing `np.empty(0)` implies to me that `0` will be the value in the array (notwithstanding the name `empty`), not that there's a shape of `0`. So I think it would be better to use `np.empty(shape=(0,))`, which is then just `self._shape`, and putting in `(0,)` I thought would be more confusing than using `self._shape`, because the question arises, if these are equal why not just use the existing variable name?

Reply:
I think explicitly showing the size as zero (with whatever notation you like) is preferable because it makes it clear that this doesn't result in an array with uninitialized values. Tracing the logic back to see that `self._shape` is `(0,)` if we're in this branch is not really that simple.

Reply:
OK, this is fixed now with `shape=(0,)` 👍